plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/catalog/snowflake/SnowflakeIcebergMetadata.java

```
@@ -0,0 +1,296 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.iceberg.catalog.snowflake;
+
+import io.airlift.json.JsonCodec;
+import io.airlift.slice.Slice;
+import io.trino.plugin.iceberg.CommitTaskData;
+import io.trino.plugin.iceberg.IcebergFileSystemFactory;
+import io.trino.plugin.iceberg.IcebergMetadata;
+import io.trino.plugin.iceberg.TableStatisticsWriter;
+import io.trino.plugin.iceberg.catalog.TrinoCatalog;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.CatalogHandle;
+import io.trino.spi.connector.CatalogSchemaTableName;
+import io.trino.spi.connector.ColumnHandle;
+import io.trino.spi.connector.ColumnMetadata;
+import io.trino.spi.connector.ConnectorInsertTableHandle;
+import io.trino.spi.connector.ConnectorMaterializedViewDefinition;
+import io.trino.spi.connector.ConnectorMergeTableHandle;
+import io.trino.spi.connector.ConnectorOutputMetadata;
+import io.trino.spi.connector.ConnectorOutputTableHandle;
+import io.trino.spi.connector.ConnectorSession;
+import io.trino.spi.connector.ConnectorTableExecuteHandle;
+import io.trino.spi.connector.ConnectorTableHandle;
+import io.trino.spi.connector.ConnectorTableLayout;
+import io.trino.spi.connector.ConnectorTableMetadata;
+import io.trino.spi.connector.ConnectorViewDefinition;
+import io.trino.spi.connector.MaterializedViewFreshness;
+import io.trino.spi.connector.RetryMode;
+import io.trino.spi.connector.SaveMode;
+import io.trino.spi.connector.SchemaTableName;
+import io.trino.spi.security.TrinoPrincipal;
+import io.trino.spi.statistics.ComputedStatistics;
+import io.trino.spi.type.TypeManager;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+
+public class SnowflakeIcebergMetadata
```

Overriding `IcebergMetadata` is not the right way to do this. Instead, this should be done in `TrinoSnowflakeCatalog`. For example, `TrinoRestCatalog` doesn't support materialized views, so those methods will throw.

---

How should there be implemented the `INSERT` (beginInsert/finishInsert) scenario in such a manner that it fails early before attempting to write data files on the object store?
I went through `TrinoGlueCatalog` & `GlueIcebergTableOperations` and I find this rather tricky.

---

Haven't changed it as I agree with @findinpath. Adding snowflake dedicated IcebergMetadata helps to early abort the unsupported operations when a method in `IcebergMetadata` performs any prior action before making a call to `TrinoSnowflakeCatalog`. Please advise if you think otherwise.

---

It should fail when the Iceberg transaction is created.

---

Removed SnowflakeIcebergMetadata and added `TrinoCatalog#newTransaction` to create an Iceberg Transaction. David, please see if it aligns with what you suggested.